kubeflow/notebooks/prototypes/README.md

```diff
@@ -0,0 +1,131 @@
+
+- allow similar params as kubeflow/core/prototypes/jupyterhub.jsonnet, including PVCs
+
+
+## Design
+
+The Notebooks component is an alternative to jupyterhub. The component defines a Notebook CRD and provides a notebook-controller (based on metacontroller's [CompositeController](https://metacontroller.app/api/compositecontroller/)). The Notebook CRD schema is similar to a kubernetes [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#creating-a-deployment) where it defines a [Pod Template](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template) that is used by the notebook-controller to create a Pod. An example Notebook yaml is shown below:
+
+```yaml
+apiVersion: kubeflow.org/v1alpha1
+kind: Notebook
+metadata:
+  name: training
+  namespace: resnet50
+spec:
+  template:
+    spec:
+      containers:
+      - name: notebook
+        image: gcr.io/kubeflow-images-public/tensorflow-1.10.1-notebook-cpu:v0.3.0
+        resources:
+          requests:
+            cpu: 500m
+            memory: 1Gi
+        workingDir: "/home/jovyan"
+      ttlSecondsAfterFinished: 300
+      securityContext: 
+      - fsGroup: 100
+        runAsUser: 1000
+```
+
+### User Interaction
+
+The user submits a Notebook yaml either through a UI or CLI (eg `kubectl apply -f notebook.yaml`) and the Notebook yaml is handled by the notebook-controller. The controller will create a Service and Pod within the user's namespace. The Service uses ambassador to create a reverse proxy that will route subsequent browser requests to the Pod. An example Service is shown below:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  annotations:
+    getambassador.io/config: |-
+      ---
+      apiVersion: ambassador/v0
+      kind:  Mapping
+      name: resnet50_training_mapping
+      prefix: /resnet50/training
+      rewrite: /resnet50/training
+      timeout_ms: 300000
+      service: training.resnet50
+  labels:
+    controller-uid: 038a2a66-e3e1-11e8-b1ac-42010a8a01b5
+  name: training
+  namespace: resnet50
+  ownerReferences:
+  - apiVersion: kubeflow.org/v1alpha1
+    blockOwnerDeletion: true
+    controller: true
+    kind: Notebook
+    name: training
+    uid: 038a2a66-e3e1-11e8-b1ac-42010a8a01b5
+spec:
+  clusterIP: 10.103.254.68
+  ports:
+  - port: 80
+    protocol: TCP
+    targetPort: 8888
+  selector:
+    app: training
+  type: ClusterIP
+```
+
+Subsequent browser requests to `https://<api-server>/<namespace>/<notebook>` are routed to the Service and Pod as shown below:
+
+![Jupyter Notebook](./docs/jupyter_notebook.png "Jupyter Notebook")
+
+The notebook component depends on the [profiles component](https://github.com/kubeflow/kubeflow/tree/master/kubeflow/profiles/prototypes) to provide a protected namespace for the user. This means that a Profile yaml was posted to the kubernetes api-server prior to posting the Notebook yaml. In this case a Profile yaml would look like:
```

Comment:
similar to @jlewi's question, is this a strict dependency? If my k8s client has permissions to create/modify namespaces, the CRD should just work right?

Reply:
yes, correct - there shouldn't be any explicit dependency on profiles. i'll review any assumptions I've made in this regard. I should also add a test case for this.